### PR TITLE
Issue-1296: set compiler path correctly for Cuda device and ensure full path for COMPILER.

### PR DESCRIPTION
**Pull request by dalabre**
This is a Work-in-Progress and should have further testing; looking for any feedback, comments, suggestions, etc.

As primary fix was for Cuda, testing only performed on White; Makefiles are created and unit tests performed and pass.

### CODE REVIEW

**Comment by dalabre** [comment]
Re-testing the travis -ci failure on kokkos-dev with module clang/4.0.1 (loads gcc/5.3.0) reproduces the error. In compilation for  example/fixture 
'''
make[2]: Entering directory `/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/example/fixture'
g++  -I./ -I/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/install/include -I/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/install/include -I/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/install/include -I/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/install/include/eti -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized --std=c++11 --gcc-toolchain=/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/5.3.0/base -I/home/dalabre/dalForkKokkos/kokkos/example/common -I/home/dalabre/dalForkKokkos/kokkos/example/fixture -O3  -c /home/dalabre/dalForkKokkos/kokkos/example/fixture/Main.cpp
g++: error: unrecognized command line option ?--gcc-toolchain=/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/5.3.0/base?
make[2]: *** [Main.o] Error 1
make[2]: Leaving directory `/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/example/fixture'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/dalabre/KWorkArea/Makefile-nvcc-Issue1296/test-on-KokkosDev/example/fixture'
make: *** [build-test] Error 2

This option comes from Makefile.kokkos:
```
# Explicitly set the GCC Toolchain for Clang.
ifeq ($(KOKKOS_INTERNAL_COMPILER_CLANG), 1)
  KOKKOS_INTERNAL_GCC_PATH = $(shell which g++)
  KOKKOS_INTERNAL_GCC_TOOLCHAIN = $(KOKKOS_INTERNAL_GCC_PATH:/bin/g++=)
  KOKKOS_CXXFLAGS += --gcc-toolchain=$(KOKKOS_INTERNAL_GCC_TOOLCHAIN)
  KOKKOS_LDFLAGS += --gcc-toolchain=$(KOKKOS_INTERNAL_GCC_TOOLCHAIN)
endif
```
Does this need tweaking?
'''

**Comment by dalabre** [comment]
The build on kokkos-dev with clang and gcc is doomed to fail as there is no Cuda device on which to build and test relative to those features in the benchmarks. When building and testing on White with the proper devpack module ( devpack/openmpi/1.10.4/ibm/xl/13.1.4/cuda/8.0.44 ) and then requesting a compute node via "bsub .....", the build and test both complete normally (no compile errors nor test failures).

Possibly the build options for the travis-ci/pr an be examined and see if the continuous integration test would pass.

**Comment by ibaned** [comment]
@dalabre can you run this spot check script with your changes?
```bash
cd kokkos
mkdir testing
cd testing
../config/test_all_sandia --spot-check
```

**Comment by dalabre** [comment]
Run spot check on white, bsub to rhel7F:   _../config/test-all_sandia  --spot-check_ , all tests FAIL:
```
/home/dalabre/kokkos/Makefile.kokkos:832: *** Makefile.kokkos: CUDA is enabled but the compiler is neither NVCC nor Clang.  Stop.
```

**Comment by ndellingwood** [comment]
@dalabre It seems from the message that `nvcc_wrapper` isn't properly being set as the compiler. One thing you can try is to run your modified `generate_makefile` and provide the `nvcc_wrapper` path as an argument (instead of letting the new changes set it automatically), this would replicate the way the Makefile is generated by the `test_all_sandia` script and if the error occurs in this instance it will be easier to triage and fix.

**Comment by ndellingwood** [comment]
Example execution:
```
# Load modules
# bsub node
cd kokkos
cd testing
../generate_makefile.bash --compiler=${PWD}/../bin/nvcc_wrapper --arch="Power8,Kepler37" --with-cuda --with-serial
```

File: generate_makefile.bash [hunk 1]
```diff
@@ -204,7 +209,10 @@ KOKKOS_SRC_PATH=${KOKKOS_PATH}
 KOKKOS_SETTINGS="KOKKOS_SRC_PATH=${KOKKOS_SRC_PATH}"
 #KOKKOS_SETTINGS="KOKKOS_PATH=${KOKKOS_PATH}"
 
-if [ ${#COMPILER} -gt 0 ]; then
+if [ ${#COMPILER} -gt 0 ] && [ ${#KOKKOS_DEVICES} -gt 0 ]; then
```
Comment (ndellingwood) [inline generate_makefile.bash]: @dalabre from our PR review, change this `if` to an `elif` and check that the `COMPILER` string is empty in that elif check, not greater than 0;
re-include the original check as the "if" case before this, i.e.
```
if [ ${#COMPILER} -gt 0 ]; then
  KOKKOS_SETTINGS="${KOKKOS_SETTINGS} CXX=${COMPILER}"
```

You may also need to set `COMPILER=""` at the start of the script, but check without doing that first. 

It's getting close :)

Comment (dalabre) [reply generate_makefile.bash]: @ndellingwood  Corrected implementation; testing thus far shows correct behavior.
Testing continues ( have a couple dozen cases to look at ); will post updated pull request tomorrow.

**Comment by dalabre** [comment]
I am going to back out this pull request. The spot check on White last night showed all Pass on Serial and OpenMP backends but the two Cuda tests (Cuda-Serial and Cuda-OpenMP) fail. I will look into the cause of failure and submit an updated pull request after tests are clean.

**Pull request closed without merging**